00-programming-fundamentals/intro-to-regex.md

```diff
@@ -117,40 +122,179 @@ Practical Example:
 
-[Sample Solutions](https://gist.github.com/CheezItMan/54e39aa4a355aa5a24bb281d5d360121)
+```ruby
+email_regex = /.+@(.+\..+)/
+match_result = "dan@adadev.org".match(email_regex)
+```
+
+We have two ways to get access to the capture groups. The first is to index into the value returned by `.match`:
+
+```ruby
+puts match_result[1]
+# => adadev.org
+```
+
+Note that capture groups start at index 1. For historical reasons index 0 stores the part of the string that matched the pattern.
+
+Your other option is to use the special variables named `$1`, `$2`, etc. We'll see this technique used when we talk about using regex with Atom below.
+
+Now that we've got access to the captured data, we can make use of it. Here is a slightly more complex version of the same program:
+
+```ruby
+emails = ['dan@adadev.org', 'dee@adadev.org', 'karib@gmail.com']
+email_regex = /.*@(.*\..*)/
 
+# Create a new hash where missing values are initialized to 0
+domain_counts = Hash.new(0)
+
+emails.each do |email|
+  match_result = email.match(email_regex)
+  next unless match_result # skip strings that don't match
+  domain = match_result[1]
+  domain_counts[domain] += 1
+end
+
+domain_counts.each do |domain, count|
+  puts "#{domain}: #{count}"
+end
+```
+
+### Practice
+1. Imagine that you have a long list of phone numbers typed by users, like the following:
+
+    ```ruby
+    phone_numbers = [
+      "(206) 555-1234",
+      "425-555-9999",
+      "406.555.1818",
+      "+1 206 555 8888",
+      "4255558872",
+    ]
+    ```
+
+    These phone numbers are formatted very differently: some have the `+1` country code, some have parentheses around the area code, some have dots instead of dashes, etc. Our job is to write some Ruby code to normalize these numbers, so that they all look like `(206) 555-1234`.
+
+    **Questions:**
+    - What regex will you need to match all these numbers?
+    - What characters will you need to capture? How many groups do you need?
+    - How will you use the captured data to fill in the final string?
+
+    [Check you answer here](solutions/regex.md#Phone%20Numbers)
+
+1. Remember back to the original RideShare project from the first week of class. We provided a [CSV file full of ride information](https://raw.githubusercontent.com/AdaGold/ride-share/master/rides.csv), and you had to manually build up a data structure of nested arrays and hashes to manage that data.
+
+    Manually copying data is a tedious, error-prone process. It would be much better if we could use our tools to do this transformation automatically. Fortunately Atom's regex search-and-replace tool is perfectly suited to this task.
+
+    We will build a regex search-and-replace pattern to transform CSV data like this
+
+    ```csv
+    DR0004,3rd Feb 2016,5,RD0022,5
+    ```
+
+    into a Ruby hashes like this
+
+    ```ruby
+    {
+      driver_id: "DR0004",
+      date: "3rd Feb 2016",
+      cost: 5,
+      rider_id: "RD0022",
+      rating: 5,
+    },
+    ```
+
```

Comment:
A lot of students are going to have trouble following these directions and translating them into text in atom.  Maybe start with a simpler example and give screenshots first, then have them do this exercise.   A picture linking capture groups and blocks of text to areas in the replace field  might help.

It's just hard to read with newlines etc and will be hard for them to correlate.